### PR TITLE
[Security] deprecate the RoleInterface

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/Security/Guard/Token/PostAuthenticationGuardToken.php [hunk 1]
```diff
@@ -28,9 +28,9 @@ class PostAuthenticationGuardToken extends AbstractToken implements GuardTokenIn
     private $providerKey;
 
     /**
-     * @param UserInterface            $user        The user!
-     * @param string                   $providerKey The provider (firewall) key
-     * @param RoleInterface[]|string[] $roles       An array of roles
+     * @param UserInterface   $user        The user!
+     * @param string          $providerKey The provider (firewall) key
+     * @param Role[]|string[] $roles       An array of roles
```
Comment (yceruto) [inline src/Symfony/Component/Security/Guard/Token/PostAuthenticationGuardToken.php]: same as before (Role|string)[] ?

Comment (xabbuh) [inline src/Symfony/Component/Security/Guard/Token/PostAuthenticationGuardToken.php]: Of course, good catch @yceruto!

**Comment by ogizanagi** [comment]
What about triggering a deprecation at least from the `SecurityDataCollector` in case a role does implement `RoleInterface` without extending `Role`?

**Comment by xabbuh** [comment]
@ogizanagi But such a role would have triggered the deprecation already because of the `@deprecated` annotation in the `RoleInterface` (at least in the `dev` environment).

**Comment by ogizanagi** [comment]
Right. I forgot about this behavior.

File: src/Symfony/Component/Security/Core/Role/RoleInterface.php [hunk 2]
```diff
@@ -18,6 +18,8 @@
  * supported by at least one AccessDecisionManager.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated The RoleInterface is deprecated since version 3.3 and will be removed in 4.0. Use the Symfony\Component\Security\Core\Role\Role class instead.
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Role/RoleInterface.php]: `Extend the ...` instead of `Use the ...`?

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by stof** [comment]
@fabpot we talked about deprecating the usage of objects as roles entirely in the past, relying on strings only. Should we move forward on this topic ?

**Comment by fabpot** [comment]
@stof Indeed, that would be a fantastic move for 4.0.

**Comment by IMM0rtalis** [comment]
> @fabpot we talked about deprecating the usage of objects as roles entirely in the past, relying on strings only. Should we move forward on this topic ?

Where was this discussed? Can you pls link this, so others and me can understand why this is the _better_ way?

**Pull request merged**